vmm/src/cpu.rs

```diff
@@ -823,6 +875,107 @@ impl CpuManager {
+
+        vcpu.lock()
+            .unwrap()
+            .fd
+            .set_cpuid2(&cpuid)
+            .map_err(Error::SetSupportedCpusFailed)?;
+
+        vcpu.lock()
+            .unwrap()
+            .restore(snapshot)
+            .expect("Failed to restore vCPU");
+
+        let vcpu_clone = Arc::clone(&vcpu);
+        self.vcpus.push(vcpu_clone);
+
+        let handle = Some(
+            thread::Builder::new()
+                .name(format!("vcpu{}", cpu_id))
+                .spawn(move || {
+                    extern "C" fn handle_signal(_: i32, _: *mut siginfo_t, _: *mut c_void) {}
+                    // This uses an async signal safe handler to kill the vcpu handles.
+                    register_signal_handler(SIGRTMIN(), handle_signal)
+                        .expect("Failed to register vcpu signal handler");
+
+                    // Block until all CPUs are ready.
+                    vcpu_thread_barrier.wait();
+
+                    loop {
+                        // vcpu.run() returns false on a KVM_EXIT_SHUTDOWN (triple-fault) so trigger a reset
+                        match vcpu.lock().unwrap().run() {
+                            Err(e) => {
+                                error!("VCPU generated error: {:?}", e);
+                                break;
+                            }
+                            Ok(true) => {}
+                            Ok(false) => {
+                                reset_evt.write(1).unwrap();
+                                break;
+                            }
+                        }
+
+                        // We've been told to terminate
+                        if vcpu_kill_signalled.load(Ordering::SeqCst)
+                            || vcpu_kill.load(Ordering::SeqCst)
+                        {
+                            break;
+                        }
+
+                        // If we are being told to pause, we park the thread
+                        // until the pause boolean is toggled.
+                        // The resume operation is responsible for toggling
+                        // the boolean and unpark the thread.
+                        // We enter a loop because park() could spuriously
+                        // return. We will then park() again unless the
+                        // pause boolean has been toggled.
+                        while vcpu_pause_signalled.load(Ordering::SeqCst) {
+                            thread::park();
+                        }
+                    }
+                })
+                .map_err(Error::VcpuSpawn)?,
+        );
+
+        // On hot plug calls into this function entry_point is None. It is for
+        // those hotplug CPU additions that we need to set the inserting flag.
+        self.vcpu_states[usize::from(cpu_id)].handle = handle;
+        self.vcpu_states[usize::from(cpu_id)].inserting = false;
+
+        Ok(())
+    }
+
```

Comment:
I have to blame myself for that one, and this was part of my TODO list, but there's a lot of common code between `restore_vcpu()` and the main `activate_vcpus()` loop. We have to factorize the vcpu thread creation code at least.

Reply:
Ok I'm gonna look at this!

Reply:
Done!